### PR TITLE
[gh] Integration tests workflow

### PR DESCRIPTION
**Pull request by boschmitt**
### Description
This PR adds a workflow to run integration tests nightly.

I had to modify some of the tests to add assertions as a mean for checking for correctness (We might want to drop the FileCheck `CHECK` for those cases, for now, I left them there). I also moved some of the tests to a new folder `NVQPP/integration`, this is to facilitate globing files that we know we want to submit to third-party simulator backends.

Since IONQ does not have measurement operations, I added a preprocessor flag `IONQ_TARGET` to handle the cases where we will submit to their simulator. More changes will be required to handle IONQ.

### CODE REVIEW

File: .github/workflows/integration_tests.yml [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+        run: |
+          curl -X POST -H "Content Type: application/json" -d '{ "email":"${{ secrets.BACKEND_LOGIN_EMAIL }}","password":"${{ secrets.QUANTINUUM_PASSWORD }}" }' https://qapi.quantinuum.com/v1/login > credentials.json
+          id_token=`cat credentials.json | jq -r '."id-token"'`
+          refresh_token=`cat credentials.json | jq -r '."refresh-token"'`
+          echo "key: $id_token" > ~/.quantinuum_config
+          echo "refresh: $refresh_token" >> ~/quantinuum_config
+
+      - name: Build and run
+        run: |
+          for filename in test/NVQPP/integration/*.cpp; do
```
Comment (anthony-santana) [inline .github/workflows/integration_tests.yml:41]: We just merged in some hardware examples that aren't tested by the CI yet ([docs/sphinx/examples/*/providers](https://github.com/NVIDIA/cuda-quantum/blob/main/docs/sphinx/examples/cpp/providers/quantinuum.cpp)). It might be worth testing that single quantinuum file and, now or eventually, running the [python](https://github.com/NVIDIA/cuda-quantum/blob/main/docs/sphinx/examples/python/providers/quantinuum.py) example as well

**Comment by schweitzpgi** [comment]
We might want to rethink these end-to-end tests a bit and come up with a better structure. They are a bit misplaced in the regression `test/` directory and now that they will have to be customized for a variety of different emulators, we will probably want to write different tests for these targets. i.e., one emulator having more functionality than the next one, taking different profiles, etc. Maybe something like a `nvq++-testsuite/` directory with different subfolders?

File: test/NVQPP/integration/bug_qubit.cpp [hunk 2]
```diff
@@ -11,27 +11,31 @@
...
 #include <iostream>
 
-struct ak2 {
+struct simple_x {
   void operator()() __qpu__ {
     cudaq::qubit q;
-    h(q);
+    x(q);
+
+// FIXME: This doesn't really depend on us. IONQ's doesn't accept measurements.
```
Comment (bettinaheim) [inline test/NVQPP/integration/bug_qubit.cpp:22]: There are things that we can and should do on our end as well. I filed an issue to track that here: https://github.com/NVIDIA/cuda-quantum/issues/512
Please link that issue instead in all of these comments.
```suggestion
// TODO: Extend measurement support for submissions to IonQ, 
// see https://github.com/NVIDIA/cuda-quantum/issues/512.
```

File: .github/workflows/integration_tests.yml [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+            [ -e "$filename" ] || continue
```
[the remark this answers is not in the document]
Comment (boschmitt) [reply .github/workflows/integration_tests.yml:42]: Not redundant. In bash, when one tries to iterate over files by globbing (which is the case), it's good practice to avoid the corner case where the glob does not match. The problem is that when the glob doesn't match, loop variable expand to the (un-matching) glob pattern string itself, and thus we would try to compile `test/NVQPP/integration/*.cpp`

[the remark this answers is not in the document]
Comment (schweitzpgi) [reply .github/workflows/integration_tests.yml:42]: Sidebar random note:

An alternative to using globbing could be something like
```sh
for filename in `find test/NVQPP/integration -maxdepth 1 -name '*.cpp'`; do
```
(Not tried under the .yml context here, so YMMV.)

Not sure which, if either, is more appealing.

File: test/NVQPP/integration/bug_qubit.cpp [hunk 4]
```diff
@@ -11,27 +11,31 @@
...
 // MLIR:           %[[VAL_0:.*]] = quake.alloca !quake.veq<1>
 // MLIR-NEXT:      %[[VAL_1:.*]] = quake.extract_ref %[[VAL_0]][0] : (!quake.veq<1>) -> !quake.ref
 
 int main() {
-  auto counts = cudaq::sample(ak2{});
-  std::cout << "Test: ";
-  counts.dump();
+  auto result = cudaq::sample(simple_x{});
+  std::cout << result.most_probable() << '\n';
+  assert("1" == result.most_probable());
```
[the remark this answers is not in the document]
Comment (boschmitt) [reply test/NVQPP/integration/bug_qubit.cpp:37]: Calling `nvq++` without the appropriate flags won't produce a release build. I suppose you are thinking about CMake's build type configuration which, when building in release, will add the appropriate flags to the compiler calls, e.g., `-O3 -DNDEBUG`, and thus disable assertions.

File: .github/workflows/integration_tests.yml [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+    runs-on: ubuntu-latest
+    container:
+      image: ghcr.io/${{ github.repository }}:latest
+      options: --user root
+
+    steps:
+      - name: Get code
+        uses: actions/checkout@v3
+        with:
+          fetch-depth: 2
```
[the remark this answers is not in the document]
Comment (bettinaheim) [reply .github/workflows/integration_tests.yml:21]: Yes, the latest image is generally produced by the latest/current commit on the main branch. 
The sequence of events here is imperfect; it is possible that a commit is on main but the image is not (yet) pushed. If you wanted to guarantee that the test folder matches the image you are pulling, then you could e.g. write the image sha + the github commit sha to a specific file on the bot/data branch, and then at the end of the day, look at the last line in that file, check out these, and delete the rest. If you want, I can take care of that. Let's first get the initial setup going without worrying too much about test and latest image not being in sync.

File: .github/workflows/integration_tests.yml (hunk 3 above)
[the remark this answers is not in the document]
Comment (bettinaheim) [reply .github/workflows/integration_tests.yml:42]: I would prefer to fail if we don't find any tests here. The expectation is that we run some tests, and if someone moves the folder without updating the workflow, then this workflow should fail.

File: test/NVQPP/integration/bug_qubit.cpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (bettinaheim) [reply test/NVQPP/integration/bug_qubit.cpp:37]: I am indeed. Hm, my initial thought was that maybe we should compile the non-kernel code with more optimization by default, but I see that for both clang and gcc, if no argument is given then the default is to do no optimization. It makes sense to have the same behavior.

File: .github/workflows/integration_tests.yml (hunk 3 above)
[the remark this answers is not in the document]
Comment (boschmitt) [reply .github/workflows/integration_tests.yml]: > I would prefer to fail if we don't find any tests here. The expectation is that we run some tests, and if someone moves the folder without updating the workflow, then this workflow should fail.

Agreed. I made changes so that instead of continuing, the loop will emit an error if the glob fails. This way if we move these files, we won't have the problem of always passing this workflow without actually testing anything.

File: .github/workflows/integration_tests.yml (hunk 5 above)
[the remark this answers is not in the document]
Comment (boschmitt) [reply .github/workflows/integration_tests.yml:39]: Seems quite involved process. Wouldn't be easier to inject the container a file with the commit SHA used to create it ? We could also output this info to the user on container startup, which would also improve bug reports, since we would be able to ask for the SHA inside the container itself.

File: .github/workflows/integration_tests.yml (hunk 1 above)
Comment (boschmitt) [reply .github/workflows/integration_tests.yml:66]: Thanks for bringing it up. I think it would be better to leave python tests for a future PR, tho. First, let's see how this workflow will work, next I will add IONQ backend, and then another workflow where we install wheel in a clean environment and test providers. What you think?

Comment (anthony-santana) [reply .github/workflows/integration_tests.yml:66]: Sounds good to me 👍

File: .github/workflows/integration_tests.yml (hunk 5 above)
[the remark this answers is not in the document]
Comment (bettinaheim) [reply .github/workflows/integration_tests.yml:39]: Good point. This file actually already exists. In the docker container, you can use the command
```
 cat $CUDA_QUANTUM_PATH/assets/build_info.txt | grep -o 'source-sha: \S*' | cut -d ' ' -f 2
```
to get the commit sha to check out on this repo.

File: .github/workflows/integration_tests.yml [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+        required: true
+        default: 'none'
+        type: choice
+        options:
+          - none
+          - quantinuum
+      target_machine:
+        type: string
+        required: false
+        description: 'Target machine (e.g., H1-1E).'
```
Comment (bettinaheim) [inline .github/workflows/integration_tests.yml:22]: You need to set a default here (should be the syntax checker). Otherwise the workflow below will fail if the target_machine is not specified.

File: .github/workflows/integration_tests.yml [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+    - cron: 0 3 * * *
+
+jobs:
+  integration_test:
+    name: Intergration test
+    runs-on: ubuntu-latest
+    environment: backend-validation
+    container:
+      image: ghcr.io/nvidia/cuda-quantum:latest-hpc
+      options: --user root
```
Comment (bettinaheim) [inline .github/workflows/integration_tests.yml:33]: I don't think running as root is necessary anymore.

**Review by bettinaheim — APPROVED** [review]
One minor comment, otherwise looks good.

File: .github/workflows/integration_tests.yml (hunk 7 above)
Comment (boschmitt) [reply .github/workflows/integration_tests.yml:33]: We still need it if we wan to use the `actions/checkout@v3`. As the non-hpc image, it seems this one won't start the container as root, and thus the checkout action won't have permission to create its files.
